pkg/imgutil/dockerconfigresolver/credentialsstore_test.go

```diff
@@ -177,6 +178,16 @@ func TestBrokenCredentialsStore(t *testing.T) {
 		for _, tc := range testCases {
 			t.Run(tc.description, func(t *testing.T) {
 				directory := tc.setup()
+				// See https://github.com/containerd/nerdctl/issues/3413
+				oldpwd, err := os.Getwd()
+				assert.NilError(t, err)
+				// Ignore the error, as the destination may not be a directory
+				_ = os.Chdir(directory)
+				t.Cleanup(func() {
```

Comment:
could we put `t.Cleanup` outside the loop ?

Reply:
Mmm... we do `Chdir` to whatever is returned by the specific _subtest_ `setup` routine, so, it should stay there, right?
Maybe I misunderstand your point?

Reply:
`t.Cleanup` are executed in LIFO order  after the test return, the last one will override all of them

Reply:
Put otherwise. « Each subtest should leave things as they were when they exit ».

Reply:
`t.Cleanup` funcs are called when the test and all its subtests complete; I think that having redundant `t.Cleanup` call for each subset test is useless.

what I suggest is:
````
oldpwd, err := os.Getwd()
assert.NilError(t, err)
directory := tc.setup()
_ = os.Chdir(directory)
t.Cleanup(func() {
	err = os.Chdir(oldpwd)
	assert.NilError(t, err)
})
for _, tc := range testCases {
			t.Run(tc.description, func(t *testing.T) {
				cs, err := NewCredentialsStore(directory)
				assert.ErrorIs(t, err, tc.errorNew)
				if err != nil {
					return
				}

				var af *Credentials
				af, err = cs.Retrieve(registryURL, true)
				assert.ErrorIs(t, err, tc.errorRead)

				err = cs.Store(registryURL, af)
				assert.ErrorIs(t, err, tc.errorWrite)

			})
		}
````

Reply:
How would `directory := tc.setup()` work outside of the loop that defines `tc`?

Reply:
I do not get where you are going with this.

The t.Cleanup calls are not "redundant". They are restoring things to what they were before the subtest started so that we do not pollute the next subtest (which is btw currently a recurring problem in our tests). 

This whole thing here is about avoiding the bad effects of a bug in docker, and work-arounding the fact that our testing rig is not segregating tests properly (we should be in a tempdir to start with).

Reply:
> How would directory := tc.setup() work outside of the loop that defines tc?

Sorry my pseudo code was wrong, this is what I am suggesting 

```
	      oldpwd, err := os.Getwd()
	      assert.NilError(t, err)
             // Register cleanup to reset cwd after executing testsCases
	      t.Cleanup(func() {
			err = os.Chdir(oldpwd)
			assert.NilError(t, err)
	      })
	      for _, tc := range testCases {
			t.Run(tc.description, func(t *testing.T) {
				directory := tc.setup()
				_ = os.Chdir(directory)
				cs, err := NewCredentialsStore(directory)
				assert.ErrorIs(t, err, tc.errorNew)
				if err != nil {
					return
				}

				var af *Credentials
				af, err = cs.Retrieve(registryURL, true)
				assert.ErrorIs(t, err, tc.errorRead)

				err = cs.Store(registryURL, af)
				assert.ErrorIs(t, err, tc.errorWrite)

			})
		}
```
> The t.Cleanup calls are not "redundant". They are restoring things to what they were before the subtest started so that we do not pollute the next subtest (which is btw currently a recurring problem in our tests).


nope, the `t.Cleanup` func is only registered after each test subset but they are all executed after the parent test returns (please check the def of [Cleanup](https://pkg.go.dev/testing@go1.23.1#T.Cleanup))

Reply:
Thanks a lot @fahedouch 
Appreciate the clarification a lot.

This is definitely a very interesting (and important conversation).

So,

```
func TestTest(t *testing.T) {

	t.Cleanup(func() {
		fmt.Println("After main test")
		assert.Assert(t, false)
	})

	t.Run("subtest1", func(tt *testing.T) {
		fmt.Println("Inside subtest1")
		tt.Cleanup(func() {
			fmt.Println("After subtest1")
		})
	})

	t.Run("subtest2", func(tt *testing.T) {
		fmt.Println("Inside subtest2")
		tt.Cleanup(func() {
			fmt.Println("After subtest2")
		})
	})

	fmt.Println("About to end main test")
}
```

Will output (expectedly):
```
Inside subtest1
After subtest1
Inside subtest2
After subtest2
About to end main test
After main test
```

This above is what I intended (cleanup executed after each subtest) - and how my test is actually behaving... but... the catch is...

The fact that my code is shadowing `t` makes things extra confusing to read (I will update my PR to not do that and use `tt` instead).

Reply:
Also, apologies for the delay in answering here. I was out for a few days.
Cheers.

Reply:
@fahedouch latest push does clarify intent by no longer shadowing `t` (we really need better test tooling...).

Let me know if the intent is now clearer.

Thanks again!

Reply:
thank you, it's clearer now!